File: testdata/fixtures/ruby/safe/crypto_strong.rb
# Safe: Strong cryptography with BCrypt and SecureRandom
# Expected: No findings for GTSS-CRY-001 or GTSS-CRY-010

require 'bcrypt'
require 'securerandom'

class AuthService
  BCRYPT_COST = 12

  def hash_password(password)
    BCrypt::Password.create(password, cost: BCRYPT_COST)
  end

  def verify_password(password, stored_hash)
    BCrypt::Password.new(stored_hash) == password
  end

  def generate_token
    SecureRandom.hex(32)
  end

  def generate_session_id
    SecureRandom.uuid
  end

  def generate_api_key
    SecureRandom.urlsafe_base64(48)
  end

  def generate_otp
    SecureRandom.random_number(1000000).to_s.rjust(6, '0')
  end
end

File: testdata/fixtures/python/safe/crypto_strong.py
# Safe: Strong cryptography with bcrypt and secrets module
# Expected: No findings for BATOU-CRY-001 or BATOU-CRY-009

import bcrypt
import secrets
import hashlib


def hash_password(password: str) -> str:
    salt = bcrypt.gensalt(rounds=12)
    return bcrypt.hashpw(password.encode(), salt).decode()


def verify_password(password: str, stored_hash: str) -> bool:
    return bcrypt.checkpw(password.encode(), stored_hash.encode())


def generate_token() -> str:
    return secrets.token_hex(32)


def generate_session_id() -> str:
    return secrets.token_urlsafe(48)


def generate_api_key() -> str:
    return secrets.token_urlsafe(64)


def generate_otp() -> str:
    return str(secrets.randbelow(1000000)).zfill(6)


def generate_csrf_token() -> str:
    return secrets.token_hex(16)


def hash_for_integrity(data: bytes) -> str:
    return hashlib.sha256(data).hexdigest()
